Search service as standalone app

lykhachov opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->
https://github.com/magento-commerce/api-protos/pull/2
<!--- ^ DO NOT REMOVE placeholder above ^ -->



### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/catalog-storefront#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes https://github.com/magento/catalog-storefront/issues/401 

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Code Review Checklist (*)


- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
- [ ] no backward incompatible changes

Commit:
added gitignore

Commit:
added gitignore

Commit:
init repo

Commit:
composer dependencies fix

Commit:
created custom attribute dependencies

Commit:
clean ConfigReader.php

owlsmage committed:
374: Remove index related logic from elastisearch

owlsmage committed:
374: Remove eav config from data provider

owlsmage committed:
374: Clean Elasticsearch

Commit:
remed store model

owlsmage committed:
374: Clean Elasticsearch

Commit:
Merge branch '374-search-service' into 374-search-service-catalog-remove

owlsmage committed:
374: Clean Elasticsearch

Commit:
clean ConfigReader.php

Commit:
removed Catalog and Eav module

Commit:
Merge remote-tracking branch 'origin/374-search-service-catalog-remove' into 374-search-service-catalog-remove

owlsmage committed:
374: Clean Elasticsearch

owlsmage committed:
374: Revert change in Elasticsearch6

owlsmage committed:
374: Mark files as copy in Elasticsearch6

owlsmage committed:
374: Mark files as copy in Elasticsearch6

Commit:
Moved Magento_Store to Magento_SearchStorefrontStore
Removed Magento_Directory
Cleaned Magento_SearchStorefrontAdvancedSearch

Commit:
Merge branch '374-search-service' of github.com:lykhachov/storefront-search-ce into 374-search-service

 Conflicts:
	app/code/Magento/SearchStorefrontElasticsearch/Model/Adapter/FieldMapper/Product/FieldProvider/DynamicField.php

Commit:
fix

Commit:
added "copied" comment to advancedSearch

Commit:
Moved Store Interface to Model from Api

Commit:
Removed Group.php and GroupRepository.php

Commit:
Cleaned up SearchStorefrontStore module

owlsmage committed:
374: Mark files as copy in Elasticsearch

owlsmage committed:
Merge remote-tracking branch 'origin/374-search-service' into 374-search-service

Commit:
changed module name in di

Commit:
moved di configs

Commit:
removed @api and @since

Commit:
cleaned up

Commit:
added comments to proto file

Commit:
cleaned up, removed CommandList.php

Commit:
cleaned up

Commit:
cleaned up

Commit:
cleaned up setup folder

Commit:
removed lib folder

owlsmage committed:
374: Remove setup folder and dependency on it

owlsmage committed:
374: Remove unused classes from elasticsearch

Commit:
removed Suggestions

Commit:
Merge remote-tracking branch 'origin/374-search-service' into 374-search-service

owlsmage committed:
374: Fix composer.json of modules

owlsmage committed:
374: Update composer.json and README.md of Magento_SearchStorefront

owlsmage committed:
374: remove unnecessary comments from di.xml

owlsmage committed:
374: update composer.json and README.md of Magento_SearchStorefrontConfig and Magento_SearchStorefrontSearch modules

owlsmage committed:
374: remove unused elasticsearch classes

owlsmage committed:
374: remove usage of category stub

Commit:
cleaned di.xml to run on minimum config available

Commit:
added ES7

Commit:
removed ES6

Commit:
cleaned up

Commit:
removed test file

Commit:
removed ES6

Commit:
put back some prefs

Commit:
cleaned up

Commit:
cleaned up

Commit:
changed init config

Commit:
cleaned ES di.xml

Commit:
removed config.xml stub and dependencies

Commit:
cleaned up

Commit:
cleaned up

Commit:
changed fallback for price buket in env.php

Commit:
added "copied" comment
moved some code back to ES module

Commit:
cleaned up

Commit:
cleaned up

Commit:
replaced EAV stubs with sql queries

Commit:
added config command to cli

Commit:
clean up

Commit:
added composer.json to modules

Commit:
removed unused interface

Commit:
changes to di.xml

Commit:
added descriptions to composer.json files

Commit:
changes according failed tests

Commit:
changes according failed tests

Commit:
renamed bootstrap, autoload
changed all dependencies in modules for these files
changed Magento_Grpc module to SearchStorefrontGrpc and all corresponding files

Commit:
renamed composer files

Commit:
renamed bin/magento to bin/storefront

Commit:
Merge branch '374-search-service' into search-service-as-standalone-app

Commit:
removed search_engine.xml

Commit:
Merge branch '374-search-service' into search-service-as-standalone-app

mslabko committed:
fix monolith installation

mslabko commented:
TODO:
Please, fix problems:
1. Add parameters to `bin/command storefront:search:init` command to be able configure connection to "DB" and "search-store-front"  
1. Validate empty values for init command and throw execption if no values provided
1. Remove SearchStorefrontGrpc module (I see only difference in namespace with GTPC module - we can copy-paste it as is for now, but later we will include it as a dependency). If any fix in origninal moudle is required, please describe them
1. Replace any "<prefererence>"  which override Magento\Framework claass, but instead introduce own interface OR specify own implementaion/stub with "<type ...><arguments>" in di.xml (see example in https://github.com/magento/storefront-search-ce/pull/4/commits/237b8fc81b53a34cd2e64026c805445aa7fe2579)

mslabko commented:
@magento run all tests

mslabko commented:
Failing on CI:
According to log we have overrides for magento classes, e.g.  Magento\SearchStorefrontConfig\App\Config\Type\System

Should be fixed after implementing item 4 from previous comment 

```[Progress: 496 / 991]
Enabling caches:

In FileLock.php line 101:

  [Magento\Framework\Exception\RuntimeException]
  Cannot acquire a lock.


Exception trace:
  at /var/www/html/lib/internal/Magento/Framework/Lock/Backend/FileLock.php:101
 Magento\Framework\Lock\Backend\FileLock->lock() at /var/www/html/lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php:134
 Magento\Framework\Cache\LockGuardedCacheLoader->lockedLoadData() at /var/www/html/app/code/Magento/SearchStorefrontConfig/App/Config/Type/System.php:275
 Magento\SearchStorefrontConfig\App\Config\Type\System->loadDefaultScopeData() at /var/www/html/app/code/Magento/SearchStorefrontConfig/App/Config/Type/System.php:201
 Magento\SearchStorefrontConfig\App\Config\Type\System->getWithParts() at /var/www/html/app/code/Magento/SearchStorefrontConfig/App/Config/Type/System.php:175
 Magento\SearchStorefrontConfig\App\Config\Type\System->get() at /var/www/html/lib/internal/Magento/Framework/App/Config.php:132
```

Commit:
cleared/added preferences

mslabko commented:
@magento run WebApi Tests

mslabko commented:
@magento run WebAPI Tests

mslabko commented:
Regarding problem with read system configuration:
"types" is missed for App\Config (see di below).
We can copy Magento\Config\App\Config\Type\System and adapt this class and all related classes to be able read system configuration.
But since this is useless work lets hard-code all places where we need system configuration, identify them (config paths) and move to a separate task (please, create new task).  After implementation https://github.com/magento/catalog-storefront/issues/427 we can start work new story and use configuration from search service directly

``` 
   <type name="Magento\Framework\App\Config">
        <arguments>
            <argument name="types" xsi:type="array">
                <item name="system" xsi:type="object">Magento\Config\App\Config\Type\System</item>
            </argument>
        </arguments>
    </type>
```

mslabko committed:
fix monolith installation

mslabko commented:
@magento run WebAPI Tests

mslabko committed:
fix monolith installation

mslabko commented:
@magento run all tests

mslabko commented:
Remaining work:
- check comment https://github.com/magento/storefront-search-ce/pull/4#issuecomment-729971677
- refactor: remove unused classes, configuration 
- implement manual & improved algorithm (see app/etc/storefront_search/di.xml)


Also I'm little bit concerned with current performance. On the local machine I see the following numbers:
```
INFO[0110] 192.168.224.10:51346 Ok {575ms} /magento.searchStorefrontApi.proto.Search/searchProducts 
INFO[0323] 192.168.224.10:51346 Ok {393ms} /magento.searchStorefrontApi.proto.Search/searchProducts 
INFO[0357] 192.168.224.10:51346 Ok {252ms} /magento.searchStorefrontApi.proto.Search/searchProducts 
INFO[0361] 192.168.224.10:51346 Ok {143ms} /magento.searchStorefrontApi.proto.Search/searchProducts 
INFO[0368] 192.168.224.10:51346 Ok {126ms} /magento.searchStorefrontApi.proto.Search/searchProducts 

```

If we split test and measure  whole service time (\Magento\SearchStorefront\Model\SearchService::searchProducts )
and time to ES only in \Magento\SearchStorefrontElasticsearch7\SearchAdapter\Adapter::query difference will be almost 3X:
ES query: 0.12847805023193
service:    0.35563087463379

After fixing existing issues need to investigate what is the issue with performance. Areas to check: 
- what and how many SQL queries  we do during request
- check existing mode and ability to switch to "production" with a standalone installation

Commit:
added arguments to config command, changed dependencies

Commit:
cleaned up

Commit:
changed di, added comments

Commit:
cleaned up

Commit:
put back Grpc module

mslabko commented:
@magento run all tests

Commit:
removed circular dependency

lykhachov commented:
@magento run all tests

lykhachov commented:
@magento run Functional Tests CE

Commit:
run /vendor/bin/php-cs-fixer

lykhachov commented:
@magento run Static Tests

Commit:
static tests

Commit:
added composer blacklist
moved files to other module

lykhachov commented:
@magento run Static Tests

lykhachov commented:
@magento run Static Tests

Commit:
undeclared tests

Commit:
statics test resolve

lykhachov commented:
@magento run all tests

Commit:
static test resolving

Commit:
static test resolving

lykhachov commented:
@magento run all tests

Commit:
remove empty line

Commit:
static tests

lykhachov commented:
@magento run all tests

Commit:
Added synonyms support

Commit:
add suppress warnings

Commit:
added comments

vgoncharenko commented on app/code/Magento/SearchStorefront/DataProvider/Product/LayeredNavigation/Builder/Category.php:
```diff
@@ -0,0 +1,151 @@
+use Magento\SearchStorefront\DataProvider\CategoryAttributesMapper;
+use Magento\SearchStorefront\DataProvider\Product\LayeredNavigation\Formatter\LayerFormatter;
+use Magento\SearchStorefront\DataProvider\Product\LayeredNavigation\LayerBuilderInterface;
+use Magento\SearchStorefrontStore\Model\Store;
```
it's better to use interface instead of particular implementation, moreover if you already have the interface.

mslabko replied:
@lykhachov, if we don't need introduced interfaces and don't use them in codebase, please remove them at all.

lykhachov replied:
replaced with Interface

mslabko commented on app/code/Magento/SearchStorefrontElasticsearch/Model/Config.php:
```diff
@@ -0,0 +1,169 @@
+    ) {
+        $this->scopeConfig = $scopeConfig;
+        $this->clientResolver = $clientResolver;
+        $this->engineResolver = $engineResolver;
```
Property is used only in constructor, perhaps we are dealing with dead code here. 
Please, check other files as well and remove any not used code (Php Inspections 
 or similar plugin can help with it, see https://github.com/kalessil/phpinspectionsea )

        $this->clientResolver = $clientResolver;
        $this->engineResolver = $engineResolver;
        $this->prefix = $prefix ?: $this->clientResolver->getCurrentEngine();
        $this->engineList = $engineList;

lykhachov replied:
removed unused code

mslabko committed:
fix static

mslabko commented:
@magento run Static Tests

mslabko committed:
add ignore list temporary modules

mslabko commented:
@magento run Static Tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

Commit:
changed direct Store dependency with Interface
replaced Arguments with Options in Install command
added make_standalone_app.sh
updated README.md

Commit:
static tests

mslabko commented:
@magento run Static Tests

Commit:
config installer change

Commit:
keep auth.json from overwriting

lykhachov commented:
@magento run all tests

mslabko committed:
fix readme, remove extra files

mslabko commented:
@magento run all tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

Commit:
static tests
removed unused

Commit:
Merge branch 'search-service-as-standalone-app' into search-storefront-synonyms

# Conflicts:
#	app/code/Magento/SearchStorefrontSearch/Model/SynonymReader.php

Commit:
Merge branch 'search-service-as-standalone-app' into search-storefront-synonyms

# Conflicts:
#	app/code/Magento/SearchStorefrontSearch/Model/SynonymReader.php

lykhachov commented:
@magento run all tests

Commit:
static tests

Commit:
static tests

Commit:
Merge branch 'search-service-as-standalone-app' into search-storefront-synonyms

lykhachov commented:
@magento run all tests

Commit:
static tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

Commit:
static tests

lykhachov commented:
@magento run Static Tests

lykhachov commented:
@magento run all tests

Commit:
static tests

Commit:
Merge branch 'search-storefront-synonyms' into search-service-as-standalone-app

lykhachov commented:
@magento run Static Tests

mslabko committed:
fix tests, remote wront dependency

mslabko commented:
@magento run all tests

mslabko committed:
fix static test

mslabko commented:
@magento run Static Tests, Semantic Version Checker, Magento Health Index

Commit:
added Magento\Elasticsearch to undeclared blacklist

lykhachov commented:
@magento run Static Tests

mslabko commented:
Semantic Version Checker and  Magento Health Index approved
"getFacets" in "obsolete" check approved

mslabko committed:
add cache config, remove extra query

RuslanKostiv1 commented:
@magento import pr to magento-commerce/storefront-search-ce

magento-engcom-team commented:
@RuslanKostiv1 the pull request successfully imported.

RuslanKostiv1 commented:
@magento run all tests

RuslanKostiv1 commented:
@magento run all tests
